### PR TITLE
Functional programming additions

### PR DESCRIPTION
**Pull request by azhyun**
This isn't ready to merge yet, but i just wanted to post my progress for the time being. Would love feedback!

Also, I'm used to creating PR's by comparing across branches. Let me know if you'd rather that I fork instead.

### CODE REVIEW

**Comment by alodahl** [comment]
Hi Alice, this format is fine!  I'll review this tonight or tomorrow and give you feedback 😊

File: functional-programming/FP.md [hunk 1]
```diff
@@ -4,10 +4,13 @@
 30-45 minutes
 
 ### Prerequisites
-
+JavaScript basics
```
Comment (alodahl) [inline functional-programming/FP.md]: could you add these lessons as markdown links?
https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet

File: functional-programming/FP.md [hunk 2]
```diff
@@ -4,10 +4,13 @@
...
+`for` loops
+arrays
+Object-oriented programming
```
Comment (alodahl) [inline functional-programming/FP.md]: FYI, object oriented programming is covered in this lesson, so no need to go into detail about it here:
https://github.com/Techtonica/curriculum/blob/master/javascript/javascript-7-oop.md

File: functional-programming/FP.md [hunk 3]
```diff
@@ -16,26 +19,42 @@
 - mutation
 - map
 - filter
+- reduce
```
Comment (alodahl) [inline functional-programming/FP.md]: 👏🏼

File: functional-programming/FP.md [hunk 4]
```diff
@@ -16,26 +19,42 @@
...
 
 ### Materials
 
 -[What is Functional Programming[(https://medium.com/javascript-scene/master-the-javascript-interview-what-is-functional-programming-7f218c68b3a0)
+-[Map, Filter, and Reduce in JavaScript (https://medium.com/jsguru/javascript-functional-programming-map-filter-and-reduce-846ff9ba492d)]
 - <img width="608" alt="steven_luscher_on_twitter___map_filter_reduce_in_a_tweet__map__ __ __ ___cook______ __ __ __filter__ __ __ ___isvegetarian______ __ __reduce__ __ ___eat_____ _" src="https://cloud.githubusercontent.com/assets/16581819/22812830/e2d723fe-eefb-11e6-87ed-775e2b5fd105.png">
 - [This other website](otherexample.com)
 
 ### Lesson
+Here are a list of tools that become available with functional programming. They are common functions that take apply functions across all elements in a container, such as an array.
```
Comment (alodahl) [inline functional-programming/FP.md]: This second sentence is confusing =/

File: functional-programming/FP.md [hunk 5]
```diff
@@ -16,26 +19,42 @@
...
+The purpose of `map` is to process every element of an array in the exact same way. When you call `map` on an array, it executes that callback on every element within it, returning a new array with all of the values that the callback returned.
+Example: array of integers, add '3' to every integer
+    var intArray = [1, 2, 3]
 
     function add3(value) {
       return value + 3;
     }
 
-    array.map(add3)
+    array.intArray(add3)
```
Comment (alodahl) [inline functional-programming/FP.md]: Can you show them the output too?  Also, I'm afraid I don't know what line 43 means, and are you still using map?

File: functional-programming/FP.md [hunk 6]
```diff
@@ -16,26 +19,42 @@
...
+The purpose of `reduce` is to process elements of an array in a way that every element will be incorporated into the final solution. When you call `reduce` on an array and a function, it applies the function to every element of the array, while keeping track of a previous value.
+Example: You are going to binge watch Season 2 of "Stranger Things", and you want to know how long it will take. You have a class called `Episode`, which has `duration` as the value, and you have all of the episodes of of Stranger Things stored in an array. 
+    var seasonTwoEpisodes = allEspisodes.filter((episode) => episode.season == 1)
+
+    function addDuration(episode1, episode2) {
+      return episode1.duration + episode2.duration;
+    }
 
-Build on the first information. Have participants guess things, do an activity, etc.
+    var total = seasonTwoEpisodes.reduce(addDuration, 0)
```
Comment (alodahl) [inline functional-programming/FP.md]: 👏

**Comment by philihp** [comment]
Can you use `let = ` instead of `var = `? Explaining why they should use `let` might be out of scope (🤓) for the lesson, but avoiding `var` entirely, at least while they're learning FP, would avoid some weird and unexpected errors.

**Comment by alodahl** [comment]
> Can you use `let = ` instead of `var = `? Explaining why they should use `let` might be out of scope (🤓) for the lesson, but avoiding `var` entirely, at least while they're learning FP, would avoid some weird and unexpected errors.

@philihp , I'm not sure we have an explanation of let v const v var anywhere.... how would you feel about adding a slide about it in [JS1](https://docs.google.com/presentation/d/1WIm5UCQL9TOsmW5X6suapBhyEqtk8Y2oLKb5gUqbzus/edit?usp=sharing)?

File: functional-programming/FP.md (hunk 2 above)
Comment (alodahl) [reply functional-programming/FP.md]: no problem, I just wanted to make sure you didn't do more work than you had to!

File: functional-programming/FP.md (hunk 1 above)
Comment (alodahl) [reply functional-programming/FP.md]: Yes please, they should be linked to the other lessons in this github Techtonica/curriculum repo.

**Comment by alodahl** [comment]
@jdq22 was able to find some great resources that would work as a 'check for understanding'.  Are you ok with this?

**Comment by ch-alicehyun** [comment]
@alodahl yes, of course that's okay to add those resources! Thanks for finding those @jdq22 :)

**Pull request merged**